pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -111,63 +112,144 @@ func (pl *VolumeZone) Filter(ctx context.Context, _ *framework.CycleState, pod *
-			if !volumeZoneLabels.Has(k) {
-				continue
-			}
-			nodeV := nodeConstraints[k]
-			volumeVSet, err := volumehelpers.LabelZonesToSet(v)
-			if err != nil {
-				klog.InfoS("Failed to parse label, ignoring the label", "label", fmt.Sprintf("%s:%s", k, v), "err", err)
-				continue
+		for _, key := range topologyLabels {
+			if value, ok := pv.ObjectMeta.Labels[key]; ok {
+				volumeVSet, err := volumehelpers.LabelZonesToSet(value)
+				if err != nil {
+					klog.InfoS("Failed to parse label, ignoring the label", "label", fmt.Sprintf("%s:%s", key, value), "err", err)
+					continue
+				}
+				podPVTopologies = append(podPVTopologies, pvTopology{
+					pvName: pv.Name,
+					key:    key,
+					values: volumeVSet,
+				})
 			}
+		}
+	}
+	return podPVTopologies, nil
+}
 
-			if !volumeVSet.Has(nodeV) {
-				klog.V(10).InfoS("Won't schedule pod onto node due to volume (mismatch on label key)", "pod", klog.KObj(pod), "node", klog.KObj(node), "PV", klog.KRef("", pvName), "PVLabelKey", k)
-				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonConflict)
-			}
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *VolumeZone) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter invoked at the filter extension point.
+//
+// It evaluates if a pod can fit due to the volumes it requests, given
+// that some volumes may have zone scheduling constraints.  The requirement is that any
+// volume zone-labels must match the equivalent zone-labels on the node.  It is OK for
+// the node to have more zone-label constraints (for example, a hypothetical replicated
+// volume might allow region-wide access)
+//
+// Currently this is only supported with PersistentVolumeClaims, and looks to the labels
+// only on the bound PersistentVolume.
+//
+// Working with volumes declared inline in the pod specification (i.e. not
+// using a PersistentVolume) is likely to be harder, as it would require
+// determining the zone of a volume during scheduling, and that is likely to
+// require calling out to the cloud provider.  It seems that we are moving away
+// from inline volume declarations anyway.
+func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	// If a pod doesn't have any volume attached to it, the predicate will always be true.
+	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	if len(pod.Spec.Volumes) == 0 {
+		return nil
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	var podPVTopologies []pvTopology
+
+	state, err := getStateData(cs)
+	if err != nil {
+		// Fallback to calculate pv list here
+		var status *framework.Status
+		podPVTopologies, status = pl.getPVbyPod(ctx, pod)
+		if !status.IsSuccess() {
+			return status
+		}
+	} else {
+		podPVTopologies = state.podPVTopologies
+	}
```

Comment:
let's return immediately if `podPVTopologies` is empty. This is possible because a Pod might have ConfigMap/Secret volumes, but not PVs.

Reply:
ConfigMaps and Secrets are still declared as Volumes and don't have PVs. But you are right... this already short-circuits